File: Library/src/main/java/org/endeavourhealth/library/model/imq/Delete.java
package org.endeavourhealth.library.model.imq;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyOrder;
import java.util.List;

@JsonPropertyOrder({ "id", "iri", "alias", "case", "aggregate", "where", "orderBy", "direction", "limit", "groupBy", "having", "select" })
@JsonInclude(JsonInclude.Include.NON_DEFAULT)
public class Delete {

  private Where property;
  private Node subject;
  private boolean inverse;
  private Node predicate;
  private Node object;
  private List<Delete> delete;

  public Where getProperty() {
    return property;
  }

  public Delete setProperty(Where where) {
    this.property = where;
    return this;
  }

  public Node getSubject() {
    return subject;
  }

  public Delete setSubject(Node subject) {
    this.subject = subject;
    return this;
  }

  public boolean isInverse() {
    return inverse;
  }

  public Delete setInverse(boolean inverse) {
    this.inverse = inverse;
    return this;
  }

  public Node getPredicate() {
    return predicate;
  }

  public Delete setPredicate(Node predicate) {
    this.predicate = predicate;
    return this;
  }

  public Node getObject() {
    return object;
  }

  public Delete setObject(Node object) {
    this.object = object;
    return this;
  }

  public List<Delete> getDelete() {
    return delete;
  }

  public Delete setDelete(List<Delete> delete) {
    this.delete = delete;
    return this;
  }
}

File: src/interfaces/Delete.ts
import z from "zod";

import { Node, NodeSchema } from "./Node";
import { Where, WhereSchema } from "./Where";

// export interface Delete {
//   property?: Where;
//   subject?: Node;
//   inverse?: boolean;
//   predicate?: Node;
//   object?: Node;
//   delete?: Delete[];
// }

export const DeleteSchema = z.strictObject({
  property: WhereSchema.optional(),
  subject: NodeSchema.optional(),
  inverse: z.boolean().default(false),
  predicate: NodeSchema.optional(),
  object: NodeSchema.optional(),
  get delete() {
    return z.array(DeleteSchema).optional();
  }
});

export type Delete = z.output<typeof DeleteSchema>;
